### ISSUE TITLE
Certain RedHat types are not available for JSON marshal/unmarshal

### ISSUE THREAD

**Issue by simar7** [aquasecurity/vuln-list-update#59]
I was using some of the types defined in the RedHat package and noticed this:
https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L23-L25

Any reason why they aren't available to be used? The JSON tag of `-` simply ignores them. 

Can we make them usable by adding their own appropriate tags?

**Comment by knqyf263** [comment]
Marshaling them fails due to two possible types.
https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L36-L38

**Comment by simar7** [comment]
> Marshaling them fails due to two possible types.
> 
> https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L36-L38

So does this mean we cannot use them? I didn't catch what you wanted to say.

**Comment by knqyf263** [comment]
You did a similar thing, so I think you can understand if you read the source code. In the following case, it is possible CVSSScore is `string` or `float64`.
https://github.com/aquasecurity/trivy-db/pull/38

Those fields might be different types depending on an advisory. Please read the comment.
https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L36-L38

Therefore, type assertion is necessary to unmarshal them.
https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L48-L64

In short, they are unmarshaled in `UnmarshalJSON` manually and should be available. Please read the source code.
https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L54
https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L78
https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L94

**Comment by simar7** [comment]
> You did a similar thing, so I think you can understand if you read the source code. In the following case, it is possible CVSSScore is `string` or `float64`.
> [aquasecurity/trivy-db#38](https://github.com/aquasecurity/trivy-db/pull/38)
> 
> Those fields might be different types depending on an advisory. Please read the comment.
> 
> https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L36-L38
> 
> Therefore, type assertion is necessary to unmarshal them.
> 
> https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L48-L64
> 
> In short, they are unmarshaled in `UnmarshalJSON` manually and should be available. Please read the source code.
> 
> https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L54
> 
> 
> https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L78
> 
> 
> https://github.com/aquasecurity/vuln-list-update/blob/ade2d21f6e7e2c757e87847e71471fe0e5caa20a/redhat/types.go#L94

Thanks for the detailed reply. I had misunderstood earlier. Now it makes sense to me. Actually it was probably something wrong on my end which was preventing the unmarshaling to occur properly. All good now!

**Comment by simar7** [comment]
This is a short example to demonstrate what I was running into.

```
package main

import (
	"encoding/json"
	"github.com/davecgh/go-spew/spew"
	"log"

	"github.com/aquasecurity/vuln-list-update/redhat"
)

type RedHatCVEJSON redhat.RedhatCVEJSON

const CVEJSON = `{
  "affected_release": [
    {
      "product_name": "Red Hat Enterprise Linux 3",
      "release_date": "2009-08-18T00:00:00Z",
      "advisory": "RHSA-2009:1218",
      "cpe": "cpe:/o:redhat:enterprise_linux:3",
      "package": "pidgin-1.5.1-4.el3"
    }
  ]
}
`

func main() {
			//got := &redhat.RedhatCVEJSON{} // good
			got := &RedHatCVEJSON{} // bad
			err := json.Unmarshal([]byte(CVEJSON), got)
			if err != nil  {
				log.Fatalf("unknown error: %s", err)
			}

			if len(got.AffectedRelease) <= 0 {
				log.Fatal("unmarshal failure")
			} else {
				log.Print("unmarshal success")
				spew.Dump(got.AffectedRelease)
			}
}
```

Maybe it's a misunderstanding on my part when I use the custom type. But when I do use a custom type, the custom` UnmarshalJSON` func for `RedHatCVEJSON` isn't called as expected. It works fine if I don't create a custom type.

**Comment by knqyf263** [comment]
Named type doesn't inherit methods. [Type assertion](https://play.golang.org/p/BYUvMhVXaAb) or [Type embdding](https://play.golang.org/p/TXT03RI79Pb) works as you expected.
